azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_mqtt.py

```diff
@@ -45,89 +69,47 @@ def _run_op(self, op):
-            #
-            # All calls into self.transport can raise an exception, and this is OK.
-            # The exception handler in PipelineStage.run_op() will catch these errors
-            # and propagate them to the caller.  This is an intentional design of the
-            # pipeline, that stages, etc, don't need to worry about catching exceptions
-            # except for special cases.
-            #
-            # The code right below this comment is This is a special case.  In addition
-            # to this "normal" exception handling, we add another exception handler
-            # into this class' Connect, Reconnect, and Disconnect code.  We need to
-            # do this because transport.on_mqtt_connected and transport.on_mqtt_disconnected
-            # are both _handler_ functions instead of _callbacks_.
-            #
-            # Because they're handlers instead of callbacks, we need to change the
-            # handlers while the connection is established.  We do this so we can
-            # know when the protocol is connected so we can move on to the next step.
-            # Once the connection is established, we change the handler back to its
-            # old value before finishing.
-            #
-            # The exception handling below is to reset the handler back to its original
-            # value in the case where transport.connect raises an exception.  Again,
-            # this extra exception handling is only necessary in the Connect, Disconnect,
-            # and Reconnect case because they're the only cases that use handlers instead
-            # of callbacks.
-            #
-            self.transport.on_mqtt_connected = on_connected
+            self._cancel_active_connect_disconnect_ops()
+            self._active_connect_op = op
             try:
                 self.transport.connect(password=self.sas_token)
             except Exception as e:
-                self.transport.on_mqtt_connected = self.on_connected
+                self._active_connect_op = None
                 raise e
 
         elif isinstance(op, pipeline_ops_base.ReconnectOperation):
             logger.info("{}({}): reconnecting".format(self.name, op.name))
 
-            @pipeline_thread.invoke_on_pipeline_thread_nowait
-            def on_connected():
-                logger.info("{}({}): on_connected.  completing op.".format(self.name, op.name))
-                self.transport.on_mqtt_connected = self.on_connected
-                self.on_connected()
-                operation_flow.complete_op(self, op)
-
-            # See "A note on exception handling" above
-            self.transport.on_mqtt_connected = on_connected
+            # We set _active_connect_op here because a reconnect is the same as a connect for "active operation" tracking purposes.
+            self._cancel_active_connect_disconnect_ops()
+            self._active_connect_op = op
```

Comment:
So, I have a slight concern about this.

We ALWAYS must call `_cancel_active_connect_disconnect_ops()` before setting the `_active_connect_op` (or the `_active_disconnect_op` for that matter). But it's not enforced. If someone were to modify this file, and forget that prerequisite call to `_cancel_active_connect_disconnect_ops()`, we would enter a land of undefined and unexpected behaviour.

In order to protect ourselves, I would ideally like to see rigorous tests for connect/disconnect/reconnect that show that we always have the expected behaviour (setting a new active op, clears all previously active ops). Another thing that could help is enforcing the cancel before the set.